qutip/semidefinite.py

```diff
@@ -184,15 +185,66 @@ def dnorm_problem(dim):
 
     logger.debug("Using {} constraints.".format(len(constraints)))
 
+    return X, constraints
+
+
+@memoize
+def dnorm_problem(dim):
+
+    X, constraints = initialize_constraints_on_dnorm_problem(dim)
```

Comment:
You're applying the memoisation to the wrong bit, here.  The reason behind memoising an operation is to cache something that takes a very long time to calculate, but will be used many many times.

As you pointed out, the complete optimisation problem itself is dependent on the matrices used - _really_, the dense problem depends on the matrices too and it's only by a little bit of trickery the previous code got around that.  You couldn't do the same thing with your sparse matrices, because it also depended on the structure.

However, if you drill down a little, you'll find that there's only actually one slow part of the whole `dnorm_problem` - most of it is very fast.  To gain all the advantages of memoisation, we only need to cache the slow bit.  If you run it in a profile with a dimension of ~100, you should find the slow part is entirely in the `qudit_swap` function, which is only used when initialising the constraints.  So, now you've split it up, you can memoise `initialize_constraints_...`, and you'll get the speed boost for both the dense and the sparse forms.

Reply:
I am  a little bit lost here, `initialize_constraints_...` has the `memoize` decorator applied already (see line 154).  Keeping another one on `dnorm_problem`  makes that implementation even "faster" than the sparse  one which already enjoys the fact that the call to
`initialize_constraints_...`  is memoized already.

Reply:
Oh, I'm sorry - I somehow overlooked that you'd applied it to `initialize_constraints`.  You were exactly right to do that.

In general I'd avoid memoisation except in case where's it's really _really_ necessary.  It has the potential to cause all sorts of problems if the output is mutable (which in Python it usually is), which are absolutely horrible to debug.  It also adds extra memory usage, though in this case it wouldn't be too expensive.

For examples of what can go wrong, consider this code that memoises when it doesn't need to:
```python
>>> @memoize
... def initial_values(x):
...     return list(range(x))
...
>>> first = initial_values(5)
>>> first
[0, 1, 2, 3, 4]
>>> second = initial_values(5)
>>> second
[0, 1, 2, 3, 4]
>>> first.append(5)
>>> first
[0, 1, 2, 3, 4, 5]
>>> second
[0, 1, 2, 3, 4, 5]
>>> initial_values(5)
[0, 1, 2, 3, 4, 5]
```
It's not really obvious why this happens at the best of times, let alone when it causes some crazy bug 10 files deep in your programme.  (There are ways around this with some Python objects, but it's still extra cognitive load and easy to mistake.)

It's a handy optimisation to know about, but optimisations almost always come at a cost, whether it's in code reusability, readability or stability.  Here, the time taken in evaluating a `dnorm` will be hugely dominated by actually solving the semidefinite problem (once the slow, dense tensor swap is memoised), so you gain no measurable advantage beyond it, and you only get the negatives.

Reply:
Neat example, thank you.